proposals/LayerProtocolRevision.md

```diff
@@ -0,0 +1,172 @@
+      let logits = model.applied(to: images)
+      let batchLoss = softmaxCrossEntropy(logits: logits, labels: labels)
+      totalLoss += batchLoss.scalarized()
+      return batchLoss
+    }
+    optimizer.update(&model.allDifferentiableVariables, along: gradients)
+  }
+  // test
+  testQueue.async(group: testGroup) {
+    var totalCorrect: Int32 = 0
+    for i in 0..<Int32(10) {
+      let images = minibatch(testImages, size: testBatchSize, batchIndex: i)
+      let labels = minibatch(numericTestLabels, size: testBatchSize, batchIndex: i)
+      let predictions = model.infer(from: images)
+      let correct = predictions.argmax(squeezingAxis: 1) .== labels
+      totalCorrect += Tensor<Int32>(correct).sum().scalarized()
+    }
+
+    let accuracy = Float(totalCorrect) / Float(numericTestLabels.shape[0])
+    print("epoch \(epoch) accuracy: \(accuracy) loss: \(totalLoss)")
+  }
+}
+testGroup.wait()
+print("Training complete: \(String(timeInterval: Date().timeIntervalSince(start)))")
+```
+## Copying Layers and the LearningPhaseIndicator
+Some operators such as BatchNorm and Dropout need to behave differently depending on whether they are performing training or inference. The current design defines the LearningPhaseIndicator class which is intended to behave like a global variable scoped to a single model. The training loop would toggle the _.training_ value depending on whether training or inference is being performed.
+
+The examples I saw in [tensorflow swift-models](https://github.com/tensorflow/swift-models) had the LearningPhaseIndicator declared and manipulated separately from the model it was affecting. One object having a side effect on another is problematic. Declaring it as a member of the root layer would have been better. In any case this design won’t work because as soon as you copy a model, the same LearningPhaseIndicator will be affecting both models. This would make it impossible to perform concurrent testing, or work with model copies in general. I don’t believe there is any clean way to have a pseudo global variable scoped to a Layer tree. I ran into the same design problem several years ago.
+
+## Suggested Design Change
+A simple and performative solution is to modify the Layer _applied(to:_ API to include a training parameter. Perhaps add an extension function to drop the parameter when calling layers that don’t make a distinction.
+```swift
+public protocol Layer: Differentiable & KeyPathIterable
+    where AllDifferentiableVariables: KeyPathIterable {
+    /// The input type of the layer.
+    associatedtype Input: Differentiable
+    /// The output type of the layer.
+    associatedtype Output: Differentiable
+
+    /// Returns the output obtained from applying to an input.
+    @differentiable(wrt: (self, input))
+    func applied(to input: Input, training: Bool) -> Output
+}
+
+public extension Layer {
+  func applied(to input: Input) -> Output {
+    return applied(to: input, training: false)
```

Comment:
I don't think it could lead to unexpected behavior make this method available. For instance, you could pass `batchNorm.applied(to: x)` without a second argument in a model's `applied(to:)` method, but it could be misleading and cause unintended behavior even though it looks okay. Moreover, semantically, sublayers' learning phase should be training, when the parent `applied(to:)` is training, so passing `false` would be semantically inconsistent.

Reply:
+1 to making the context parameter required (and not defining this extension). It'd probably be a code smell to not pass through the context / generate a new context implicitly. Thanks!